geopandas/tools/sjoin.py

```diff
@@ -6,6 +6,9 @@
 from geopandas.array import _check_crs, _crs_mismatch_warn
 
 
+ALLOWED_OP = {"intersects", "contains", "within"}
```

Comment:
@adriangb it seems incongruous that `ALLOWED_OP` is here, whereas `allowed_hows` is within `_basic_checks` below.  (seems like things that perform a similar role should be in a similar place)

Even if ops are added in later refactors (e.g., based on new predicates from `query_bulk`), it seems like this could remain inline within `_basic_checks` unless you foresee needing to check `ALLOWED_OPS` in other places (in which case maybe that is the time to promote to a global variable?)

Reply:
My hope is that in the future we offload this check/task to `sindex.query_bulk` since that is what really determines what `op` values are valid.* I felt that making it a global variable like this could help with that goal, but I do see your point. I'll move it back if nothing else to minimize the diff.

* I know this is off topic, but to just to start thinking about: aside from `within/contains` there is no `op/predicate` logic in `sjoin` anymore, so do we need tests for all of the other predicates before adding them? Or are we okay just having tests for `within`, `contains` and one of the other predicates (maybe `intersects` because tests are already implemented)? The rest of the testing would be up to `sindex` or, further down the pipeline, `pygeos`/`shapely2`.

Reply:
Following up on this: perhaps the best option here is to not check the `op` parameter and let it be caught by `sindex` and bubble up. In the future, when `sindex` is user-defined and/or on a per-column basis, or also if we give users the ability to plug their own `_geom_predicate_query`, we would have to do it that way anyhow.

Reply:
>  perhaps the best option here is to not check the op parameter and let it be caught by sindex and bubble up

That sounds reasonable. Do we get a meaningful/compatible error message there?

Reply:
In the case of pygeos, an invalid predicate (`foo`) raises:

```
ValueError: Predicate foo is not valid; must be one of within, crosses, contains, touches, overlaps, intersects
```
(other predicates will be added to list here as they are supported in pygeos)

Reply:
> perhaps the best option here is to not check the op parameter and let it be caught by sindex and bubble up

This seems reasonable to me too.

> do we need tests for all of the other predicates before adding them?

I'm not sure about the `rtree` side, but we're striving for good unit test coverage of the predicates in pygeos - but we're still a bit weak on this across all geometry types.

The more I thought about this, the more I wanted the `sindex` tests in geopandas to be a bit more like integration tests.  Or put a bit differently, the unit tests in pygeos are using contrived geometries that have more easily known geometric relationships.  In geopandas, it might be good to verify at least some of these with geometries derived from real-world datasets, which would require a bit more manual review of the test outputs.

All that said, I think it is reasonable to rely on the core libraries to have good unit tests ("trust, but verify"), and add more thorough `sindex` tests in later PRs.

Reply:
Totally agree @brendan-ward!

Regarding the error raised, we are currently checking the predicates in `sjoin` and a `ValueError` is raised for invalid predicates like so:
```
Got `predicate` = `foo`, `predicate` must be one of {None, within, crosses, contains, touches, overlaps, intersects}"
```
I think that is a reasonable error to see from `sjoin`. It may be a duplication of what `pygeos` is doing, but if we were to support more spatial index implementations we would want a unified interface, so I think having this duplication makes sense.

I'll implement this change tonight and update back here. Thanks all!

Reply:
> having this duplication makes sense

agreed - thanks for following up on that, I didn't have a chance to check the latest in geopandas.

Reply:
I realized that by removing this check, we would technically allow users to use other predicates, ex: `crosses`. I _expect_ it to work fine since it's just passing through to `sindex`, but there are no tests for this behavior. Do we need `sjoin` specific tests for this?

Reply:
I'm not sure I follow.  If the predicates are validated at the `sindex` level, and they don't change the nature in which `sjoin` operates, then I don't think we need additional tests at the `sjoin` level to validate that those predicates are in the allowed list.

Reply:
Exactly. That's my thought. I just want to make sure everyone is okay with that. 

To clarify: currently we have tests for `within`, `contains` and `intersects` in `sjoin`. `sjoin` _does_ have some logic regarding the `op` parameter, but since 3868b44e1a087565ba16ff228966f757bb58a7de it is limited to flipping `within/contains` or doing nothing otherwise. So I think it makes sense to keep the existing tests, but I do not think we need to add new ones. If we all agree on this (it sounds like you do @brendan-ward) we should go ahead and enable the rest of the predicates by dropping the check in `sjoin` and relying on `sindex` to decide what it can and cannot do.

The reason I think this refactor is the place to make this change is to avoid introducing the new `_basic_checks` function with `op` in it's signature only to change that a couple of commits later.

Reply:
Okay seems like there is consensus! @jorisvandenbossche please chime in real quick if you can.

What I'll do is:
1. Make a new PR to adjust the `sjoin` documentation to reflect this change and assign it to myself.
2. Push the change to this PR to implement.